File: lib/services/executors/claude-agent/src/cli-runner.ts
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.apache at the repo root.

import { spawn, ChildProcess } from "node:child_process";
import { writeFile, unlink, mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { buildCliEnv, type CliAuthConfig } from "./cli-env.js";
import { CALLBACK_MCP_SERVER_NAME } from "./internal-mcp-server.js";
import { TOOL_DEFINITIONS } from "./internal-mcp-tools.js";

export const REQUIRED_CALLBACK_TOOLS: string[] = TOOL_DEFINITIONS.map(
  (t) => `mcp__${CALLBACK_MCP_SERVER_NAME}__${t.name}`,
);

export function buildAllowedTools(templateAllowed?: string[]): string[] {
  const merged = [...REQUIRED_CALLBACK_TOOLS, ...(templateAllowed ?? [])];
  return [...new Set(merged)];
}

export type CliToolConfig =
  | {
      kind: "mcp-http";
      name: string;
      url: string;
      headers?: Record<string, string>;
    }
  | {
      kind: "mcp-stdio";
      name: string;
      command: string;
      args?: string[];
      env?: Record<string, string>;
    };

export interface CliSpawnRequest {
  model: string;
  systemPrompt: string;
  userPrompt: string;
  tools: CliToolConfig[];
  env: Record<string, string>;
  cwd?: string;
  sessionId?: string;
  bare?: boolean;
  permissionMode?: string;
  allowedTools?: string[];
  disallowedTools?: string[];
  addDirs?: string[];
  maxBudgetUsd?: string;
}

export interface CliResumeRequest {
  sessionId: string;
  prompt: string;
  tools: CliToolConfig[];
  allowedTools?: string[];
  env: Record<string, string>;
  cwd?: string;
}

export interface CliHandle {
  readonly pid?: number;
  onStdout(cb: (chunk: string) => void): void;
  onStderr(cb: (chunk: string) => void): void;
  onExit(
    cb: (exitCode: number | null, signal: NodeJS.Signals | null) => void,
  ): void;
  sendSigterm(): void;
  sendSigkill(): void;
  waitExit(): Promise<{
    exitCode: number | null;
    signal: NodeJS.Signals | null;
  }>;
}

export interface CliRunner {
  spawn(req: CliSpawnRequest): Promise<CliHandle>;
  resume?(req: CliResumeRequest): Promise<CliHandle>;
}

export function buildClaudeCliArgs(
  req: CliSpawnRequest,
  paths: { systemPromptPath: string; mcpConfigPath: string },
): string[] {
  const permissionMode = req.permissionMode ?? "bypassPermissions";
  const maxBudgetUsd = req.maxBudgetUsd ?? process.env.RIMSKY_DISPATCH_MAX_USD;
  return [
    "--print",
    "--output-format",
    "stream-json",
    "--verbose",
    "--model",
    req.model,
    "--permission-mode",
    permissionMode,
    ...(req.sessionId ? ["--session-id", req.sessionId] : []),
    ...(req.bare ? ["--bare"] : []),
    "--allowedTools",
    buildAllowedTools(req.allowedTools).join(" "),
    ...(req.disallowedTools && req.disallowedTools.length > 0
      ? ["--disallowedTools", req.disallowedTools.join(" ")]
      : []),
    ...(req.addDirs && req.addDirs.length > 0
      ? ["--add-dir", ...req.addDirs]
      : []),
    "--system-prompt-file",
    paths.systemPromptPath,
    "--mcp-config",
    paths.mcpConfigPath,
    ...(maxBudgetUsd ? ["--max-budget-usd", maxBudgetUsd] : []),
    "-p",
    req.userPrompt,
  ];
}

export function buildClaudeCliResumeArgs(
  req: CliResumeRequest,
  paths: { mcpConfigPath: string },
): string[] {
  return [
    "--resume",
    req.sessionId,
    "--print",
    "--output-format",
    "stream-json",
    "--verbose",
    "--allowedTools",
    buildAllowedTools(req.allowedTools).join(" "),
    "--mcp-config",
    paths.mcpConfigPath,
    "-p",
    req.prompt,
  ];
}

function mcpConfigJson(tools: CliToolConfig[]): string {
  const mcpServers: Record<string, unknown> = {};
  for (const t of tools) {
    if (t.kind === "mcp-http") {
      mcpServers[t.name] = {
        type: "http",
        url: t.url,
        headers: t.headers ?? {},
      };
    } else if (t.kind === "mcp-stdio") {
      mcpServers[t.name] = {
        type: "stdio",
        command: t.command,
        ...(t.args && t.args.length > 0 ? { args: t.args } : {}),
        ...(t.env ? { env: t.env } : {}),
      };
    }
  }
  return JSON.stringify({ mcpServers });
}

function buildHandleFromChild(
  child: ChildProcess,
  onCleanup: () => void,
): CliHandle {
  const stdoutCbs: ((chunk: string) => void)[] = [];
  const stderrCbs: ((chunk: string) => void)[] = [];
  const exitCbs: ((
    code: number | null,
    signal: NodeJS.Signals | null,
  ) => void)[] = [];
  type ExitResult = {
    exitCode: number | null;
    signal: NodeJS.Signals | null;
  };
  let exited = false;
  let exitResult: ExitResult | null = null;
  const exitWaiters: ((r: ExitResult) => void)[] = [];

  child.stdout?.setEncoding("utf-8");
  child.stderr?.setEncoding("utf-8");
  child.stdout?.on("data", (chunk: string) => {
    for (const cb of stdoutCbs) cb(chunk);
  });
  child.stderr?.on("data", (chunk: string) => {
    for (const cb of stderrCbs) cb(chunk);
  });

  let cleanupRan = false;
  const runCleanup = (): void => {
    if (cleanupRan) return;
    cleanupRan = true;
    try { onCleanup(); } catch {}
  };

  child.on("exit", (code, signal) => {
    exited = true;
    const result: ExitResult = { exitCode: code, signal };
    exitResult = result;
    for (const cb of exitCbs) cb(code, signal);
    for (const w of exitWaiters) w(result);
    runCleanup();
  });

  child.on("error", (err) => {
    for (const cb of stderrCbs) cb(`[spawn error] ${String(err)}\n`);
    if (!exited) {
      exited = true;
      const result: ExitResult = { exitCode: null, signal: null };
      exitResult = result;
      for (const cb of exitCbs) cb(null, null);
      for (const w of exitWaiters) w(result);
    }
    runCleanup();
  });

  return {
    pid: child.pid,
    onStdout: (cb) => { stdoutCbs.push(cb); },
    onStderr: (cb) => { stderrCbs.push(cb); },
    onExit: (cb) => { exitCbs.push(cb); },
    sendSigterm: () => { child.kill("SIGTERM"); },
    sendSigkill: () => { child.kill("SIGKILL"); },
    waitExit: () =>
      exited && exitResult
        ? Promise.resolve(exitResult)
        : new Promise((resolve) => exitWaiters.push(resolve)),
  };
}

export function createClaudeCliRunner(opts: {
  auth: CliAuthConfig;
  binaryPath?: string;
  exposeEnvNames?: readonly string[];
}): CliRunner {
  const binary = opts.binaryPath ?? "claude";
  const auth = opts.auth;
  const exposeEnvNames = opts.exposeEnvNames ?? [];
  const collectExposedEnv = (): Record<string, string> => {
    const out: Record<string, string> = {};
    for (const name of exposeEnvNames) {
      const v = process.env[name];
      if (v !== undefined) out[name] = v;
    }
    return out;
  };
  return {
    async spawn(req: CliSpawnRequest): Promise<CliHandle> {
      const tmp = await mkdtemp(join(tmpdir(), "rimsky-cli-"));
      const systemPromptPath = join(tmp, "system.md");
      const mcpConfigPath = join(tmp, "mcp.json");
      await writeFile(systemPromptPath, req.systemPrompt);

      await writeFile(mcpConfigPath, mcpConfigJson(req.tools));

      const { env: authEnv, cleanup: cleanupAuthEnv } = buildCliEnv(auth);
      const args = buildClaudeCliArgs(req, { systemPromptPath, mcpConfigPath });
      const child: ChildProcess = spawn(binary, args, {
        cwd: req.cwd,
        env: { ...collectExposedEnv(), ...authEnv, ...req.env },
        stdio: ["ignore", "pipe", "pipe"],
      });

      return buildHandleFromChild(child, () => {
        void unlink(systemPromptPath).catch(() => {});
        void unlink(mcpConfigPath).catch(() => {});
        void rm(tmp, { recursive: true, force: true }).catch(() => {});
        try { cleanupAuthEnv(); } catch {}
      });
    },
    async resume(req: CliResumeRequest): Promise<CliHandle> {
      const tmp = await mkdtemp(join(tmpdir(), "rimsky-cli-"));
      const mcpConfigPath = join(tmp, "mcp.json");
      await writeFile(mcpConfigPath, mcpConfigJson(req.tools));

      const { env: authEnv, cleanup: cleanupAuthEnv } = buildCliEnv(auth);
      const args = buildClaudeCliResumeArgs(req, { mcpConfigPath });
      const child: ChildProcess = spawn(binary, args, {
        cwd: req.cwd,
        env: { ...collectExposedEnv(), ...authEnv, ...req.env },
        stdio: ["ignore", "pipe", "pipe"],
      });
      return buildHandleFromChild(child, () => {
        void unlink(mcpConfigPath).catch(() => {});
        void rm(tmp, { recursive: true, force: true }).catch(() => {});
        try { cleanupAuthEnv(); } catch {}
      });
    },
  };
}

File: lib/services/executors/claude-agent/clirunner.go
// Copyright © 2026 Fall Guy Consulting.
// SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-FallGuy-Commercial

package claudeagent

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
)

func RequiredCallbackTools() []string {
	defs := toolDefinitions()
	out := make([]string, 0, len(defs))
	for _, d := range defs {
		out = append(out, "mcp__"+CallbackMCPServerName+"__"+d.Name)
	}
	return out
}

func BuildAllowedTools(templateAllowed []string) []string {
	merged := append(RequiredCallbackTools(), templateAllowed...)
	seen := make(map[string]bool, len(merged))
	out := make([]string, 0, len(merged))
	for _, t := range merged {
		if !seen[t] {
			seen[t] = true
			out = append(out, t)
		}
	}
	return out
}

const (
	CliToolKindMcpHTTP  = "mcp-http"
	CliToolKindMcpStdio = "mcp-stdio"
)

type CliToolConfig struct {
	Kind    string
	Name    string
	URL     string
	Headers map[string]string
	Command string
	Args    []string
	Env     map[string]string
}

type CliSpawnRequest struct {
	Model           string
	SystemPrompt    string
	UserPrompt      string
	Tools           []CliToolConfig
	Env             map[string]string
	ExposeEnvNames  []string
	Cwd             string
	SessionID       string
	Bare            bool
	PermissionMode  string
	AllowedTools    []string
	DisallowedTools []string
	AddDirs         []string
	MaxBudgetUSD    string
}

type CliResumeRequest struct {
	SessionID       string
	NewSessionID    string
	Prompt          string
	Tools           []CliToolConfig
	Model           string
	AllowedTools    []string
	DisallowedTools []string
	PermissionMode  string
	Bare            bool
	AddDirs         []string
	MaxBudgetUSD    string
	Env             map[string]string
	ExposeEnvNames  []string
	Cwd             string
}

type ExitResult struct {
	ExitCode *int
	Signal   string
}

type CliHandle interface {
	Pid() int
	OnStdout(cb func(chunk string))
	OnStderr(cb func(chunk string))
	OnExit(cb func(result ExitResult))
	SendSigterm()
	SendSigkill()
	WaitExit() ExitResult
}

type CliRunner interface {
	Spawn(req CliSpawnRequest) (CliHandle, error)
	Resume(req CliResumeRequest) (CliHandle, error)
}

type CliArgPaths struct {
	SystemPromptPath string
	McpConfigPath    string
}

type cliSessionConfig struct {
	Model           string
	PermissionMode  string
	Bare            bool
	AllowedTools    []string
	DisallowedTools []string
	AddDirs         []string
	MaxBudgetUSD    string
}

func rejectFlagLike(field, value string) error {
	if strings.HasPrefix(value, "-") {
		return &CliConfigError{Message: fmt.Sprintf(
			"cli config %s value %q begins with '-'; refusing to splice a node-supplied value that the claude CLI would parse as a flag",
			field, value)}
	}
	return nil
}

const defaultPermissionMode = "bypassPermissions"

func sessionConfigArgs(cfg cliSessionConfig) ([]string, error) {
	permissionMode := cfg.PermissionMode
	if permissionMode == "" {
		permissionMode = defaultPermissionMode
	}
	maxBudgetUSD := cfg.MaxBudgetUSD
	if maxBudgetUSD == "" {
		maxBudgetUSD = os.Getenv("RIMSKY_CLAUDE_AGENT_DISPATCH_MAX_USD")
	}
	if err := rejectFlagLike("model", cfg.Model); err != nil {
		return nil, err
	}
	if err := rejectFlagLike("permission_mode", permissionMode); err != nil {
		return nil, err
	}
	if err := rejectFlagLike("max_budget_usd", maxBudgetUSD); err != nil {
		return nil, err
	}
	for _, dir := range cfg.AddDirs {
		if err := rejectFlagLike("add_dirs", dir); err != nil {
			return nil, err
		}
	}
	args := []string{"--model", cfg.Model, "--permission-mode", permissionMode}
	if cfg.Bare {
		args = append(args, "--bare")
	}
	args = append(args, "--allowedTools", joinTools(BuildAllowedTools(cfg.AllowedTools)))
	if len(cfg.DisallowedTools) > 0 {
		args = append(args, "--disallowedTools", joinTools(cfg.DisallowedTools))
	}
	if len(cfg.AddDirs) > 0 {
		args = append(args, "--add-dir")
		args = append(args, cfg.AddDirs...)
	}
	if maxBudgetUSD != "" {
		args = append(args, "--max-budget-usd", maxBudgetUSD)
	}
	return args, nil
}

func BuildClaudeCliArgs(req CliSpawnRequest, paths CliArgPaths) ([]string, error) {
	block, err := sessionConfigArgs(cliSessionConfig{
		Model:           req.Model,
		PermissionMode:  req.PermissionMode,
		Bare:            req.Bare,
		AllowedTools:    req.AllowedTools,
		DisallowedTools: req.DisallowedTools,
		AddDirs:         req.AddDirs,
		MaxBudgetUSD:    req.MaxBudgetUSD,
	})
	if err != nil {
		return nil, err
	}
	args := []string{"--print", "--output-format", "stream-json", "--verbose"}
	if req.SessionID != "" {
		args = append(args, "--session-id", req.SessionID)
	}
	args = append(args, block...)
	args = append(args, "--system-prompt-file", paths.SystemPromptPath, "--mcp-config", paths.McpConfigPath)
	return args, nil
}

func BuildClaudeCliResumeArgs(req CliResumeRequest, paths CliArgPaths) ([]string, error) {
	block, err := sessionConfigArgs(cliSessionConfig{
		Model:           req.Model,
		PermissionMode:  req.PermissionMode,
		Bare:            req.Bare,
		AllowedTools:    req.AllowedTools,
		DisallowedTools: req.DisallowedTools,
		AddDirs:         req.AddDirs,
		MaxBudgetUSD:    req.MaxBudgetUSD,
	})
	if err != nil {
		return nil, err
	}
	args := []string{"--resume", req.SessionID, "--print", "--output-format", "stream-json", "--verbose"}
	if req.NewSessionID != "" {
		args = append(args, "--session-id", req.NewSessionID)
	}
	args = append(args, block...)
	args = append(args, "--mcp-config", paths.McpConfigPath)
	return args, nil
}

func joinTools(tools []string) string {
	return strings.Join(tools, " ")
}

func mcpConfigJSON(tools []CliToolConfig) ([]byte, error) {
	mcpServers := map[string]any{}
	for _, t := range tools {
		switch t.Kind {
		case CliToolKindMcpHTTP:
			headers := t.Headers
			if headers == nil {
				headers = map[string]string{}
			}
			mcpServers[t.Name] = map[string]any{
				"type":    "http",
				"url":     t.URL,
				"headers": headers,
			}
		case CliToolKindMcpStdio:
			entry := map[string]any{
				"type":    "stdio",
				"command": t.Command,
			}
			if len(t.Args) > 0 {
				entry["args"] = t.Args
			}
			if t.Env != nil {
				entry["env"] = t.Env
			}
			mcpServers[t.Name] = entry
		}
	}
	return json.Marshal(map[string]any{"mcpServers": mcpServers})
}

type CliRunnerOpts struct {
	Auth       CliAuthConfig
	BinaryPath string
}

type claudeCliRunner struct {
	auth   CliAuthConfig
	binary string
}

// @decision: cli-spawn-mechanism
func NewClaudeCliRunner(opts CliRunnerOpts) CliRunner {
	binary := opts.BinaryPath
	if binary == "" {
		binary = "claude"
	}
	return &claudeCliRunner{auth: opts.Auth, binary: binary}
}

func collectExposedEnv(names []string) map[string]string {
	out := map[string]string{}
	for _, name := range names {
		if v, ok := os.LookupEnv(name); ok {
			out[name] = v
		}
	}
	return out
}

func (r *claudeCliRunner) Spawn(req CliSpawnRequest) (CliHandle, error) {
	tmp, err := os.MkdirTemp("", "rimsky-cli-")
	if err != nil {
		return nil, err
	}
	systemPromptPath := filepath.Join(tmp, "system.md")
	mcpConfigPath := filepath.Join(tmp, "mcp.json")
	if err := os.WriteFile(systemPromptPath, []byte(req.SystemPrompt), 0o644); err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	mcpJSON, err := mcpConfigJSON(req.Tools)
	if err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	if err := os.WriteFile(mcpConfigPath, mcpJSON, 0o644); err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	authEnv, err := BuildCliEnv(r.auth)
	if err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	cleanup := func() {
		_ = os.RemoveAll(tmp)
		authEnv.Cleanup()
	}
	args, err := BuildClaudeCliArgs(req, CliArgPaths{
		SystemPromptPath: systemPromptPath,
		McpConfigPath:    mcpConfigPath,
	})
	if err != nil {
		cleanup()
		return nil, err
	}
	return spawnChild(r.binary, args, req.Cwd, mergeEnv(collectExposedEnv(req.ExposeEnvNames), authEnv.Env, req.Env), req.UserPrompt, cleanup)
}

func (r *claudeCliRunner) Resume(req CliResumeRequest) (CliHandle, error) {
	tmp, err := os.MkdirTemp("", "rimsky-cli-")
	if err != nil {
		return nil, err
	}
	mcpConfigPath := filepath.Join(tmp, "mcp.json")
	mcpJSON, err := mcpConfigJSON(req.Tools)
	if err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	if err := os.WriteFile(mcpConfigPath, mcpJSON, 0o644); err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	authEnv, err := BuildCliEnv(r.auth)
	if err != nil {
		_ = os.RemoveAll(tmp)
		return nil, err
	}
	cleanup := func() {
		_ = os.RemoveAll(tmp)
		authEnv.Cleanup()
	}
	args, err := BuildClaudeCliResumeArgs(req, CliArgPaths{McpConfigPath: mcpConfigPath})
	if err != nil {
		cleanup()
		return nil, err
	}
	return spawnChild(r.binary, args, req.Cwd, mergeEnv(collectExposedEnv(req.ExposeEnvNames), authEnv.Env, req.Env), req.Prompt, cleanup)
}

func mergeEnv(layers ...map[string]string) []string {
	merged := map[string]string{}
	for _, layer := range layers {
		for k, v := range layer {
			merged[k] = v
		}
	}
	out := make([]string, 0, len(merged))
	for k, v := range merged {
		out = append(out, k+"="+v)
	}
	return out
}

type realCliHandle struct {
	mu         sync.Mutex
	cmd        *exec.Cmd
	stdoutCbs  []func(string)
	stderrCbs  []func(string)
	exitCbs    []func(ExitResult)
	stdoutHist []string
	stderrHist []string
	exited     bool
	exitResult ExitResult
	done       chan struct{}
}

func spawnChild(binary string, args []string, cwd string, env []string, stdin string, cleanup func()) (CliHandle, error) {
	cmd := exec.Command(binary, args...)
	cmd.Dir = cwd
	cmd.Env = env
	cmd.Stdin = strings.NewReader(stdin)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		cleanup()
		return nil, err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		cleanup()
		return nil, err
	}
	if err := cmd.Start(); err != nil {
		cleanup()
		return nil, err
	}

	h := &realCliHandle{cmd: cmd, done: make(chan struct{})}
	var pumps sync.WaitGroup
	pumps.Add(2)
	go h.pump(stdout, &h.stdoutCbs, &h.stdoutHist, &pumps)
	go h.pump(stderr, &h.stderrCbs, &h.stderrHist, &pumps)
	go func() {
		pumps.Wait()
		err := cmd.Wait()
		result := exitResultFrom(cmd, err)
		h.mu.Lock()
		h.exited = true
		h.exitResult = result
		cbs := append([]func(ExitResult){}, h.exitCbs...)
		h.mu.Unlock()
		close(h.done)
		for _, cb := range cbs {
			cb(result)
		}
		cleanup()
	}()
	return h, nil
}

func exitResultFrom(cmd *exec.Cmd, waitErr error) ExitResult {
	state := cmd.ProcessState
	if state == nil {
		return ExitResult{}
	}
	if ws, ok := state.Sys().(syscall.WaitStatus); ok && ws.Signaled() {
		return ExitResult{Signal: ws.Signal().String()}
	}
	code := state.ExitCode()
	if code < 0 && waitErr != nil {
		return ExitResult{}
	}
	return ExitResult{ExitCode: &code}
}

func (h *realCliHandle) pump(r io.Reader, cbs *[]func(string), hist *[]string, pumps *sync.WaitGroup) {
	defer pumps.Done()
	buf := make([]byte, 32*1024)
	for {
		n, err := r.Read(buf)
		if n > 0 {
			chunk := string(buf[:n])
			h.mu.Lock()
			if len(*cbs) == 0 {
				*hist = append(*hist, chunk)
			} else {
				for _, cb := range *cbs {
					cb(chunk)
				}
			}
			h.mu.Unlock()
		}
		if err != nil {
			return
		}
	}
}

func (h *realCliHandle) Pid() int {
	if h.cmd.Process == nil {
		return 0
	}
	return h.cmd.Process.Pid
}

func (h *realCliHandle) OnStdout(cb func(string)) {
	h.registerStream(&h.stdoutCbs, &h.stdoutHist, cb)
}

func (h *realCliHandle) OnStderr(cb func(string)) {
	h.registerStream(&h.stderrCbs, &h.stderrHist, cb)
}

func (h *realCliHandle) registerStream(cbs *[]func(string), hist *[]string, cb func(string)) {
	h.mu.Lock()
	defer h.mu.Unlock()
	for _, chunk := range *hist {
		cb(chunk)
	}
	*hist = nil
	*cbs = append(*cbs, cb)
}

func (h *realCliHandle) OnExit(cb func(ExitResult)) {
	h.mu.Lock()
	if h.exited {
		result := h.exitResult
		h.mu.Unlock()
		cb(result)
		return
	}
	h.exitCbs = append(h.exitCbs, cb)
	h.mu.Unlock()
}

func (h *realCliHandle) SendSigterm() {
	if h.cmd.Process != nil {
		_ = syscall.Kill(-h.cmd.Process.Pid, syscall.SIGTERM)
	}
}

func (h *realCliHandle) SendSigkill() {
	if h.cmd.Process != nil {
		_ = syscall.Kill(-h.cmd.Process.Pid, syscall.SIGKILL)
	}
}

func (h *realCliHandle) WaitExit() ExitResult {
	<-h.done
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.exitResult
}
